public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/SmartAlertTypeDetector.tsx

```diff
@@ -0,0 +1,170 @@
+import { DataSourceJsonData } from '@grafana/schema';
+import { Alert, Button, useStyles2 } from '@grafana/ui';
+import { contextSrv } from 'app/core/core';
+import { ExpressionDatasourceUID } from 'app/features/expressions/types';
+import { AccessControlAction } from 'app/types';
+import { AlertQuery } from 'app/types/unified-alerting-dto';
+
+import { RuleFormType, RuleFormValues } from '../../../types/rule-form';
+import { NeedHelpInfo } from '../NeedHelpInfo';
+
+function getAvailableRuleTypes() {
+  const canCreateGrafanaRules = contextSrv.hasAccess(
+    AccessControlAction.AlertingRuleCreate,
+    contextSrv.hasEditPermissionInFolders
+  );
+  const canCreateCloudRules = contextSrv.hasAccess(AccessControlAction.AlertingRuleExternalWrite, contextSrv.isEditor);
+  const defaultRuleType = canCreateGrafanaRules ? RuleFormType.grafana : RuleFormType.cloudAlerting;
+
+  const enabledRuleTypes: RuleFormType[] = [];
+  if (canCreateGrafanaRules) {
+    enabledRuleTypes.push(RuleFormType.grafana);
+  }
+  if (canCreateCloudRules) {
+    enabledRuleTypes.push(RuleFormType.cloudAlerting, RuleFormType.cloudRecording);
+  }
+
+  return { enabledRuleTypes, defaultRuleType };
+}
+
+const onlyOneDSInQueries = (queries: AlertQuery[]) => {
+  return queries.filter((q) => q.datasourceUid !== ExpressionDatasourceUID).length === 1;
+};
+const getCanSwitch = ({
+  queries,
+  ruleFormType,
+  editingExistingRule,
+  rulesSourcesWithRuler,
+}: {
+  rulesSourcesWithRuler: Array<DataSourceInstanceSettings<DataSourceJsonData>>;
+  queries: AlertQuery[];
+  ruleFormType: RuleFormType | undefined;
+  editingExistingRule: boolean;
+}) => {
+  // get available rule types
+  const availableRuleTypes = getAvailableRuleTypes();
+
+  // check if we have only one query in queries and if it's a cloud datasource
+  const onlyOneDS = onlyOneDSInQueries(queries);
+  const dataSourceIdFromQueries = queries[0]?.datasourceUid ?? '';
+  const isRecordingRuleType = ruleFormType === RuleFormType.cloudRecording;
+  // it's a smart type if we are creating a new rule and it's not a recording rule type
+  const showSmartTypeSwitch = !editingExistingRule && !isRecordingRuleType;
+  //let's check if we have a smart cloud type
+  const canBeCloud =
+    showSmartTypeSwitch &&
+    onlyOneDS &&
+    rulesSourcesWithRuler.some(
+      (dsJsonData: DataSourceInstanceSettings<DataSourceJsonData>) => dsJsonData.uid === dataSourceIdFromQueries
+    );
+  // check for enabled types
+  const grafanaTypeEnabled = availableRuleTypes.enabledRuleTypes.includes(RuleFormType.grafana);
+  const cloudTypeEnabled = availableRuleTypes.enabledRuleTypes.includes(RuleFormType.cloudAlerting);
+  // can we switch to the other type? (cloud or grafana)
+  const canSwitch =
+    !editingExistingRule &&
+    !isRecordingRuleType &&
+    ((cloudTypeEnabled && canBeCloud && ruleFormType === RuleFormType.grafana) ||
+      (ruleFormType === RuleFormType.cloudAlerting && grafanaTypeEnabled));
+  return canSwitch;
```

Comment:
I wonder what we should do when the user selects the `type: range` query. AFAIK we cannot translate that into Prometheus query

Reply:
I tested and it works correctly, it ignores this type range